extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/UserService.java

```diff
@@ -460,6 +460,47 @@ public ModeledUser retrieveSkeletonUser(AuthenticationProvider authenticationPro
         return user;
         
     }
+    
+    /**
+     * Create a user in the database that does not already exist, setting up an
+     * empty model and inserting both the entity and the user object, and
+     * generating a random password for the account.
+     * 
+     * @param authenticationProvider
+     *     The authentication provider that authenticated the user.
+     * 
+     * @param authenticatedUser
+     *     The authenticated user that is being added to the database.
+     * 
+     * @return
+     *     The ModeledUser associated with the newly-created database object
+     *     for the user.
+     * 
+     * @throws GuacamoleException
+     *     If a ModeledUser cannot be created, or if the user cannot be
+     *     inserted into the database.
+     */
+    public ModeledUser createMissingUser(AuthenticationProvider authenticationProvider,
+            AuthenticatedUser authenticatedUser) throws GuacamoleException {
+        
+        ModeledUser user = getObjectInstance(null,
+                new UserModel(authenticatedUser.getIdentifier()));
+        
+        // Insert the database object
+        entityMapper.insert(user.getModel());
+            
+        // Auto-generate a password
+        user.setPassword(null);
+            
+        // Set up cyclic reference
+        user.setCurrentUser(new ModeledAuthenticatedUser(authenticatedUser,
+            authenticationProvider, user));
+            
+        // Insert the user object
+        userMapper.insert(user.getModel());
```

Comment:
The creation process will need to take into account the default set of permissions intended to be granted to each user upon creation:

https://github.com/apache/guacamole-client/blob/68c542787c092aae22decc40307b1927119f6e4b/extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/UserService.java#L73-L79

I'm a little worried about duplicating creation logic here. Is there a way to leverage `createObject()`?

Reply:
@mike-jumper: Ping - suggestions on the best way to go for this?

Reply:
Tweaking `createObject()` (and pretty much everything else) such that it has defined behavior for some unrestricted internal system user makes sense. It would also ease the implementation of some future addition to the extension API which allows third-party extensions to perform actions without requiring the logged-in user to have sufficient privileges (enrolling in TOTP despite lacking `UPDATE` permission on your own user object, for example).

Defining behavior for `null` would be simple but dangerous. It would be all to easy for a bug to result in an unexpected `null` and then bam: privilege escalation.

I like the idea of an internal system user definition. That would require very explicit usage which would not be error prone. I can think of two possible approaches:

1. Creating some wrapper class which *might* contain a `ModeledAuthenticatedUser` (but isn't guaranteed to) and which has some well-defined way to declare that an instance represents a privileged system user that lacks a model.
2. Creating some base interface which both `RemoteAuthenticatedUser` and `ModeledAuthenticatedUser` implement and which abstracts away permission retrieval. Existing usages of those classes can then be replaced with the interface, and a concrete, privileged implementation of that interface can be created which simply returns true for all permission checks.

Reply:
A third option:

Check whether it's possible to refactor things which accept `ModeledAuthenticatedUser` to instead accept `User`. Callers would then either pass `modeledAuthenticatedUser.getUser()` or the previously-mentioned concrete implementation of the interface (in this case `User`) which simply returns true for all permission checks.

Reply:
I’m getting back to working on this, hopefully have something figured out in a couple of days.  Any of the options are going to take some significant refactoring, so just trying to flesh out which makes the most sense.

Reply:
Okay, so, been taking a look at the options and trying some things out - I'm determined to figure this out...
1. The first option - a wrapper class - seems like the least desirable to me, so I've been mostly ignoring this route.  Still a possibility, but the other two seem a little more elegant.
2. The second option - an interface for permission gathering that can be implemented by `RemoteAuthenticatedUser` and `ModeledAuthenticatedUser` - seems like the best way to do this.  **Question: Is there any reason the interface to be implemented shouldn't be the `Permissions` interface, since it already contains the permissions-gathering methods?**
3. I took a look at the third option - replacing `ModeledAuthenticatedUser` with `User`, and, while it probably would be feasible in the end, it isn't quite as simple as just swapping them out - particularly since `ModeledAuthenticatedUser.getUser()` passes a `ModeledUser`, not a plain `User`, which has some extra methods like `isAdministrator()` that have to be factored in.  So, either the `User` class (interface?) would have to have at least the `isAdministrator()` method defined (stubbed), or we'd have to do `ModeledUser` instead of `User`, but then this System user couldn't be defined in the `guacamole-ext` package, it would have to be in the `guacamole-jdbc-base` package, which may not be the best place for it.  Hopefully that makes sense.
4. I guess this brings up a fourth option that is a combination of two and three - either add a handful of the permissions methods to the `User` interface, and replace `ModeledAuthenticatedUser` with `User` such that swapping out `ModeledAuthenticatedUser` with `User` would actually work.

Reply:
I'll have to reread and reacquaint myself with these changes.

I don't think it will make sense to expose permission sets from `AuthenticatedUser`, nor that there should need to be any fundamental changes to guacamole-ext. It should be possible for the internals of the JDBC auth to have their own internal representation of a system-level user without requiring that permissions be re-anchored to `AuthenticatedUser`.

There are semantic issues with doing so:

1. The `AuthenticatedUser` interface is a pure means of representing _identity_. Mixing that with identity (authentication) with permissions (authorization) collapses an established separation of concerns.
2. There is no definition of cross-extension identity except for `AuthenticatedUser`. As such, the identifiers used by `ObjectPermissionSet` will not have a well-defined meaning. Given a particular `AuthenticatedUser`, I can retrieve the corresponding `User` for each other extension and discover what extension-specific permissions have been granted. If those permissions are moved to `AuthenticatedUser`, this becomes impossible, and the meaning of the identifiers becomes ambiguous.

Reply:
Looking over creation of a user and the functionality of `createObject()`, the main beast appears to be the automatic permission creation. When a user `X` attempts to create a user `Y`, the following permissions are automatically added:

This user... | ...can do this... | ...to this object
------------ | ----------------- | -----------------
`X` | `READ` | User `Y`
`X` | `UPDATE` | User `Y`
`X` | `DELETE` | User `Y`
`X` | `ADMINISTER` | User `Y`
`Y` | `READ` | User `Y`

This will be problematic where `X` is the special system user, as then `X` does not exist and we really only want to grant the following permissions:

This user... | ...can do this... | ...to this object
------------ | ----------------- | -----------------
`Y` | `READ` | User `Y`

Perhaps we really can't avoid having a special-case `createMissingUser()`-type function specific to `UserService`...

Reply:
Okay - I'll refocus it on solely the JDBC extension for now.

I was thinking that it actually might dovetail nicely with GUACAMOLE-617, which deals with extracting the permissions management portions of the JDBC module out into a common place usable by multiple modules, but at this point it probably just makes sense to go ahead and stick in the JDBC module and work about those changes later.  Let me see what I can figure out.

Reply:
Okay, @mike-jumper, jumping back into this.  The changes you made have vastly simplified the code....

....but, I've run aground of another issue.  Your first request, long, long ago, was to try to re-use the existing `userService.createObject()` method, both to avoid duplicating code, and to make sure that implicit permissions are granted.  I'm now hitting an issue with this, because the `PrivilegedModeledAuthenticatedUser` doesn't have an `entity_id`, which means when it goes to run `getImplicitPermissions()`, it sets the `entity_id` to `null`, which then causes a database constraint violation trying to insert into `guacamole_user_permission` with a `null` `entity_id` value.

Suggestions?  Refactor `createObject()` to handle these cases?  Or scrap that and just manually do the `entityMapper.insert()` and `userMapper.insert()` calls, and then update permissions after those?

Reply:
Okay, with a couple of minor tweaks to `getImplicitPermissions()`, I think I've got this working.